### PR TITLE
create AbstractBubble, SimpleBubble class hierarchy

### PR DESCRIPTION
**Pull request by calvinnhieu**
Introduced AbstractBubble as superclass to SimpleBubble and PowerUp classes.

Changes:
- [Major] `AbstractBubble` provides abstract method definitions 
- [Major] `SimpleBubble` and `PowerUp` override and define method implementations
- [Minor] Refactored previous use cases of `Bubble` to `AbstractBubble`, and `SimpleBubble` where appropriate
- [Minor] Updated tests

*[Major] changes highlight the major changes to the code that should be focused on during code review.

### CODE REVIEW

**Comment by defnoob** [comment]
It would probably be better to just call the abstract class Bubble.

Aslo maybe move the class variables from the abstractbubble to simple bubble to get rid of the visibility issue?(Not sure if that would solve things tho.)

**Comment by calvinnhieu** [comment]
@defnoob I see your point, but I'd like to keep the naming specific - it was quite a pain refactoring uses of "Bubble". Thoughts?

The visibility issue relates to all variables (class and instance), but you're right, that might solve it - will try tonight.

**Comment by defnoob** [comment]
Well if you were to look at general frameworks and also at the example in the slides, the base/super class isn't called AbstractClass, but just Class. Have an array of AbstractBubbles is weird.

**Comment by calvinnhieu** [comment]
Kk

**Comment by calvinnhieu** [comment]
@defnoob revised 👍

File: bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java [hunk 1]
```diff
@@ -110,347 +49,154 @@
...
-
     /**
      * Returns the gamehead.
      * @return  GameData gamehead
      */
-     //CHECKSTYLE:OFF: DesignForExtension
-    public GameData getGameHead() {
-      return this.gamehead;
-    }
+    //CHECKSTYLE:OFF: DesignForExtension
```
Comment (defnoob) [inline bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java]: Can remove these lines since it's abstract now.
(Remove the at line 57(setGameHead) and line 72(draw)

File: bust-a-move-framework/src/main/java/bustamove/bubble/SimpleBubble.java [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+            default:
+                break;
+        }
+    }
+
+    /**
+     * Sets the gamehead, this is a instance of GameData.
+     * @param game the gamehead to refer to
+     */
+    //CHECKSTYLE:OFF: DesignForExtension
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/bubble/SimpleBubble.java]: I think the Checkstyle OFF/ON are not needed anymore

File: bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java [hunk 3]
```diff
@@ -28,7 +29,11 @@
     /**
      * Root bubble component.
      */
-    private Bubble rootBubble;
+    private SimpleBubble rootBubble;
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java]: Bubble instead of SimpleBubble

File: bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java [hunk 4]
```diff
@@ -262,23 +253,38 @@ public final Bubble getBubble() {
...
-
     /**
      * Setter method: sets the gamedata object that corresponds to this PowerUp.
      *
      * @param game the gamehead to refer to
      */
     public final void setGameHead(final GameData game) {
-        this.gamehead = game;
-        this.bubble.setGameHead(game);
+        // this.gamehead = game;
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java]: Remove this

File: bust-a-move-framework/src/main/java/bustamove/game/Cannon.java [hunk 5]
```diff
@@ -35,8 +36,10 @@
     /**
      * X and Y coordinate of the next bubble.
      */
-    private static final int X_LAUNCH = (int) (375 - Bubble.DIAMETER / 2);
-    private static final int Y_LAUNCH = (int) (555 - Bubble.DIAMETER / 2);
+    private static final int X_LAUNCH =
+        (int) (375 - Bubble.DIAMETER / 2);
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/game/Cannon.java]: Doesn't checkstyle complain about magic numbers ?

File: bust-a-move-framework/src/main/java/bustamove/game/Cannon.java [hunk 6]
```diff
@@ -138,7 +141,7 @@ private void loadNextBubble() {
      */
     private Bubble getNextBubble() {
         Log.log(this, "Next bubble loaded to cannon");
-        Bubble newBubble = Bubble.randomColor(
+        Bubble newBubble = SimpleBubble.randomColor(
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/game/Cannon.java]: SimpleBubble -> Bubble

File: bust-a-move-framework/src/main/java/bustamove/sound/SoundHandler.java [hunk 7]
```diff
@@ -20,7 +20,7 @@
     /**
      * The volume at which all audio will be played (between 0 and 1).
      */
-    private static float volume = 1;
+    private static float volume = 0;
```
Comment (peace-4-you) [inline bust-a-move-framework/src/main/java/bustamove/sound/SoundHandler.java]: why this change ?

File: bust-a-move-framework/src/main/java/bustamove/game/Cannon.java (hunk 6 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/game/Cannon.java]: randomColor is defined in SimpleBubble

File: bust-a-move-framework/src/main/java/bustamove/game/Cannon.java (hunk 5 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/game/Cannon.java]: it doesnt

File: bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java (hunk 4 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java]: k

File: bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java [hunk 8]
```diff
@@ -110,347 +49,154 @@
...
-        if (PowerUp.class.isAssignableFrom(this.getClass())) {
-            return this.bubble.getRootBubble();
-        } else if (this.getClass() == Bubble.class) {
-            return this;
-        } else {
-            return null;
-        }
-    }
-
+    public abstract SimpleBubble getRootBubble();
```
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java]: good catch

File: bust-a-move-framework/src/main/java/bustamove/bubble/SimpleBubble.java (hunk 2 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/bubble/SimpleBubble.java]: nice

File: bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java (hunk 3 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/bubble/powerup/PowerUp.java]: k

File: bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java (hunk 1 above)
Comment (calvinnhieu) [inline bust-a-move-framework/src/main/java/bustamove/bubble/Bubble.java]: 👍

**Comment by calvinnhieu** [comment]
Comments addressed

**Comment by defnoob** [comment]
Nice!
